R/init.R

```diff
@@ -122,52 +121,59 @@ init <- function(data,
       )
     )
   }
-
+  if (is.list(data) && !inherits(data, "teal_data_module")) {
+    checkmate::assert_list(data, names = "named")
+    data <- do.call(teal.data::teal_data, data)
+  }
   checkmate::assert_multi_class(data, c("teal_data", "teal_data_module"))
-  checkmate::assert_multi_class(modules, c("teal_module", "list", "teal_modules"))
+
+  ## `modules`
+  if (inherits(modules, "teal_module")) {
+    modules <- list(modules)
+  }
+  if (inherits(modules, "list")) {
+    modules <- do.call(teal::modules, modules)
+  }
+  checkmate::assert_class(modules, "teal_modules")
+
+  ## `filter`
   checkmate::assert(
     checkmate::check_class(filter, "teal_slices"),
     checkmate::check_list(filter, names = "named")
   )
+
+  ## all other arguments
   checkmate::assert_multi_class(title, c("shiny.tag", "character"))
   checkmate::assert_multi_class(header, c("shiny.tag", "character"))
   checkmate::assert_multi_class(footer, c("shiny.tag", "character"))
   checkmate::assert_character(id, max.len = 1, any.missing = FALSE)
 
+  # log
   teal.logger::log_system_info()
 
-  if (is.character(title)) {
-    title <- build_app_title(title)
-  } else {
-    validate_app_title_tag(title)
-  }
-
-  if (inherits(modules, "teal_module")) {
-    modules <- list(modules)
-  }
-  if (inherits(modules, "list")) {
-    modules <- do.call(teal::modules, modules)
-  }
-
+  # argument mutation
+  ## `modules` - landing module
   landing <- extract_module(modules, "teal_module_landing")
-  if (length(landing) > 1L) stop("Only one `landing_popup_module` can be used.")
+  landing_module <- NULL
+  if (length(landing) > 1L) {
+    stop("Only one `landing_popup_module` can be used.")
+  } else if (length(landing) == 1) {
+    landing_module <- landing[[1L]]
+  }
```

Comment:
I disagree with this change. You want to complete checks in one place, I get it but the check at the place of execution cannot be avoided. Your offset is you are effectively adding one more check on the result of `extract_module`, you introduce a "declare and modify" on `landing_module` here _and_ you remove a variable, which is honestly something I have never seen in a function body before.

Reply:
> but the check at the place of execution cannot be avoided.

Are you referring to this (code w/o my change)?

```
      if (length(landing) == 1L) {
        landing_module <- landing[[1L]]
        do.call(landing_module$server, c(list(id = "landing_module_shiny_id"), landing_module$server_args))
      }
```

I think it's the same as:

```
if (!is.null(landing)) { ... }
```

(or in the version with changes)
```
if (!is.null(landing_module)) { ... }
```

because we also have:

```
if (length(landing) > 1L) stop("Only one `landing_popup_module` can be used.")
```

The above practically means that we can only have 0 or 1 landing module and 0 is represented by NULL value here.

---

The whole point of my change is that I want to move a following line `landing_module <- landing[[1L]]` upper to keep all the logic together. Right now we have it a little bit distributed - _extract_ and _stop_ in one place and _use first el_ in the other.

In the light of the above - I introduced a second variable to keep interpretation of the return value from extract() next to the actual call. Currently we have `length(landing)>1` in one place and `length(landing) == 1L` together with `landing[[1]]` in the other.

---

I added a `rm()` call not to be confused during debugging which variable means what. Probably better to put better names instead. I am thinking now of having `landing_modules` (return from the extract()) and `landing_module` (as the first element of the former). WDYT?

Reply:
> > but the check at the place of execution cannot be avoided.
> 
> Are you referring to this (code w/o my change)?

Yes. You have to check down there anyway, so creating a NULL just to test for NULL instead of length>1 is redundant.


> The above practically means that we can only have 0 or 1 landing module and 0 is represented by NULL value here.

Yes. Your logic does not improve the current one, merely substitutes the length>1 condition for is.null.

> The whole point of my change is that I want to move a following line `landing_module <- landing[[1L]]` upper to keep all the logic together. Right now we have it a little bit distributed - _extract_ and _stop_ in one place and _use first el_ in the other.
> 
> In the light of the above - I introduced a second variable to keep interpretation of the return value from extract() next to the actual call. Currently we have `length(landing)>1` in one place and `length(landing) == 1L` together with `landing[[1]]` in the other.

You are restoring the logic that was proposed and rejected during review when this change was introduced.

Reply:
> so creating a NULL just to test for NULL instead of length>1 is redundant.

As pointed out in my previous message is rather to keep interpretation of extract() return value closer to the extract() function call. Let's imagine a refactor of extract() to return list() instead of NULL in case of no landing module - it would be much easier perform such refactor if both _execute_ and _use_ returned value is close together. My aim is to have pythonic `landing_module, modules <- extract_landing(modules)` where everything is in one line. Unfortunately it's not possible in R and we have to split it in multiple calls. I have no problem with that but I do have if those calls are spread across the code which I am trying to fix.

> Yes. Your logic does not improve the current one, merely substitutes the length>1 condition for is.null.

I feel that currently the information that there has to be a single landing module is implemented in two distant places in the code - my changes condense it into one if. It also moves usage of extract() return close to the call.

> You are restoring the logic that was proposed and rejected during review when this change was introduced.

Hmm... That would mean that I agree with whoever proposed this :) But if that really have been discussed than I am ok to revert it but I have to admit that I am not super convinced - just bad smell for me for future refactor.